File: java/AlgorithmAndDataStructure/src/main/java/algorithms/BinarySearch.java
package algorithms;

import java.util.List;

/**
 * Created by lzx on 15-2-20.
 * All algorithms in BinarySearch Class require a sorted List as input
 */
public class BinarySearch {
	public static <T extends Comparable> int lowerBound(List<T> a, T value) {
		int guess = a.size();
		int start = 0;
		while (guess > 0) {
			int i = start;
			int step = guess / 2;
			i += step;
			if (a.get(i).compareTo(value) < 0) {
			    start = i + 1;
				guess -= step + 1;
			}
			else {
				guess = step;
			}
		}
		return start;
	}

	public static <T extends Comparable> int upperBound(List<T> a, T value) {
		int guess = a.size();
		int start = 0;
		while (guess > 0) {
			int i = start;
			int step = guess / 2;
			i += step;
			if (a.get(i).compareTo(value) <= 0) {
				start = i + 1;
				guess -= step + 1;
			}
			else {
				guess = step;
			}
		}
		return start;
	}

	public static <T extends Comparable> boolean binarySearch(List<T> a, T
			value) {
		int i = lowerBound(a, value);
		return i != a.size() && a.get(i).equals(value);
	}

	public static <T extends Comparable> Pair<Integer, Integer> equalRange(
			List<T> a, T value) {
		return new Pair<Integer, Integer>(lowerBound(a, value), upperBound(a,
				value));
	}
}

File: scala/src/main/scala/algorithm/BinarySearch.scala
package algorithm

object BinarySearch {
  def lowerBound[T](c: Seq[T], value: T)(implicit ord: math.Ordering[T]) = {
    var guess = c.size
    var start = 0
    while (guess > 0) {
      var i = start
      i += guess / 2
      if (ord.lt(c(i), value)) {
        start = i + 1
        guess = guess / 2 - 1
      }
      else {
        guess = guess / 2
      }
    }
    start
  }

  def upperBound[T](c: Seq[T], value: T)(implicit ord: math.Ordering[T]) = {
    var guess = c.size
    var start = 0
    while (guess > 0) {
      var i = start
      i += guess / 2
      if (ord.lteq(c(i), value)) {
        start = i + 1
        guess = guess / 2 - 1
      }
      else {
        guess = guess / 2
      }
    }
    start
  }

  def binarySearch[T](c: Seq[T], value: T)(implicit ord: math.Ordering[T]) = {
    val i = lowerBound[T](c, value)
    i != c.size && ord.equiv(c(i), value)
  }

  def equalRange[T](c: Seq[T], value: T)(implicit ord: math.Ordering[T]) = {
    (lowerBound[T](c, value), upperBound[T](c, value))
  }
}
